src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokensFileViewProviderHelper.java

```diff
@@ -33,7 +33,33 @@
  */
 public class LSPSemanticTokensFileViewProviderHelper implements LSPSemanticTokensContainer {
 
+    /**
+     * Simple short-duration thread-local cache for the file view provider's PSI file to minimize read-through during
+     * burst reads of information from the file view provider.
+     */
+    private static final class CachedPsiFile {
+        // Cache a valid PSI file for up to one second
+        private static final int MAX_DURATION = 1000;
+
+        private final PsiFile file;
+        private final long created = System.currentTimeMillis();
+
+        private CachedPsiFile(@NotNull PsiFile file) {
+            this.file = file;
+        }
+
+        @Nullable
+        private PsiFile getFile() {
+            return isValid() ? file : null;
+        }
+
+        private boolean isValid() {
+            return file.isValid() && ((System.currentTimeMillis() - created) < MAX_DURATION);
+        }
+    }
+
     private final LSPSemanticTokensFileViewProvider fileViewProvider;
+    private final ThreadLocal<CachedPsiFile> cachedFilePtr = new InheritableThreadLocal<>();
```

Comment:
Oh, and this is an inheritable thread local so that the cache can be leveraged if logic moves to another thread.